lib/Sema/DerivedConformanceCodable.cpp

```diff
@@ -0,0 +1,950 @@
+
+    // If the var decl didn't validate, it may still not have a type; confirm it
+    // has a type before ensuring the type conforms to Codable.
+    if (varDecl->hasType() &&
+        tc.conformsToProtocol(varDecl->getType(), codableProto, context,
+                              ConformanceCheckFlags::Used)) {
+      callback(varDecl);
+    } else {
+      // TODO: We should produce a diagnostic note here explaining that we found
+      //       a var not conforming to Codable.
+      allConform = false;
+    }
+  }
+
+  return allConform;
+}
+
+// Validates the given CodingKeys enum decl by ensuring its cases are a 1-to-1
+// match with the stored vars of the given type.
+static bool
+validateCodingKeysEnum(TypeChecker &tc, EnumDecl *codingKeysDecl,
+                       NominalTypeDecl *type) {
+  // Look through all var decls in the given type.
+  // * Filter out lazy/computed vars (currently already done by
+  //   getStoredProperties).
+  // * Filter out ones which are present in the given decl (by name).
+  //
+  // If any of the entries in the CodingKeys decl are not present in the type
+  // by name, then this decl doesn't match.
+  // If there are any vars left in the type, then this decl doesn't match.
+  //
+  // NOTE: If we change the behavior to ignore vars with default values, then we
+  //       can further filter out the type names to remove those which
+  //       correspond to vars with default values.
+  llvm::DenseSet<Identifier> names;
+
+  auto storedProperties = type->getStoredProperties(/*skipInaccessible=*/true);
+  auto validVarCallback = [&names](VarDecl *varDecl) {
+    names.insert(varDecl->getName());
+  };
+
+  if (!validateVarsConformToCodable(tc, type->getDeclContext(),
+                                    storedProperties, validVarCallback))
+    return false;
+
+  for (auto elt : codingKeysDecl->getAllElements()) {
+    auto it = names.find(elt->getName());
+    if (it == names.end()) {
+      // TODO: Produce diagnostic here complaining that the CodingKeys enum
+      //       contains a case which does not correspond to a var.
+      // TODO: Investigate typo-correction here; perhaps the case name was
+      //       misspelled and we can provide a fix-it.
+      return false;
+    }
+
+    names.erase(it);
+  }
+
+  // TODO: Produce diagnostic here complaining that there are vars which are not
+  //       listed in the CodingKeys enum.
+  return names.empty();
+}
+
+// Fetches or synthesizes a CodingKeys type nested in the given type.
+//
+// If the given type declaration has a nested CodingKeys entity, returns it if
+// it is a valid NominalTypeDecl (nullptr otherwise).
+// If the given type declaration does not have a nested CodingKeys entity,
+// synthesizes a new enum based on the Codable members of the given type
+// (nullptr if unable to synthesize).
+static std::pair<NominalTypeDecl*, /* attempted to synthesize? */bool>
+getOrSynthesizeCodingKeysDecl(TypeChecker &tc, Decl *parentDecl,
```

Comment:
This function is huge. It clearly has several nice subroutines in it. Can you split this up?